packages/x-tree-view-pro/src/internals/plugins/useTreeViewLazyLoading/useTreeViewLazyLoading.ts

```diff
@@ -0,0 +1,296 @@
+
+      // handle caching here
+      const cachedData = cache.get(id);
+
+      if (cachedData !== undefined && cachedData !== -1) {
+        nestedDataManager.setRequestSettled(id);
+        instance.addItems({ items: cachedData, depth, parentId: id, getChildrenCount });
+        instance.setDataSourceLoading(id, false);
+
+        return;
+      }
+      if (cachedData === -1) {
+        instance.removeChildren(id);
+      }
+
+      const existingError = selectorGetTreeItemError(store.value, id) ?? null;
+      if (existingError) {
+        instance.setDataSourceError(id, null);
+      }
+
+      try {
+        const getTreeItemsResponse = await getTreeItems(id);
+        nestedDataManager.setRequestSettled(id);
+
+        // set caching
+        cache.set(id, getTreeItemsResponse);
+        // update the items in the state
+        instance.addItems({ items: getTreeItemsResponse, depth, parentId: id, getChildrenCount });
+      } catch (error) {
+        const childrenFetchError = error as Error;
+        // handle errors here
+        instance.setDataSourceError(id, childrenFetchError);
+        instance.removeChildren(id);
+      } finally {
+        // unset loading
+        instance.setDataSourceLoading(id, false);
+
+        nestedDataManager.setRequestSettled(id);
+      }
+    },
+    [nestedDataManager, params.dataSource, instance, cache, store],
+  );
+
+  useInstanceEventHandler(instance, 'beforeItemToggleExpansion', async (eventParameters) => {
+    if (!isLazyLoadingEnabled) {
+      return;
+    }
+
+    if (selectorIsItemExpanded(store.value, eventParameters.itemId)) {
+      return;
+    }
+
+    eventParameters.isExpansionPrevented = true;
+    await instance.fetchItems([eventParameters.itemId]);
+    const fetchErrors = Boolean(selectorGetTreeItemError(store.value, eventParameters.itemId));
+    if (!fetchErrors) {
+      instance.setItemExpansion(eventParameters.event, eventParameters.itemId, true);
+      if (selectorIsItemSelected(store.value, eventParameters.itemId)) {
+        // make sure selection propagation works correctly
+        instance.selectItem({
+          event: eventParameters.event as React.SyntheticEvent,
+          itemId: eventParameters.itemId,
+          keepExistingSelection: true,
+          shouldBeSelected: true,
+        });
+      }
+    }
+  });
+
+  React.useEffect(() => {
+    if (isLazyLoadingEnabled && firstRenderRef.current) {
+      if (params.items.length) {
+        const getChildrenCount = params.dataSource?.getChildrenCount || (() => 0);
+        instance.addItems({ items: params.items, depth: 0, getChildrenCount });
+      } else {
+        instance.fetchItems();
+      }
+      firstRenderRef.current = false;
+    }
+  }, [instance, params.items, params.dataSource, isLazyLoadingEnabled]);
+
+  const pluginContextValue = React.useMemo(
+    () => ({ lazyLoading: params.dataSource !== undefined }),
+    [params.dataSource],
+  );
+
+  if (!isLazyLoadingEnabled) {
```

Comment:
I think we should always return everything and inside the methods do nothing if the lazy loading is not enabled.
Otherwise, `hasPlugin` can return true and the method still be undefined which is counterintuitive and inconsistent with features like the label editing or the reordering.

You just need to make `instance.preventItemUpdates` conditional.